src/Lykke.Blockchains.Integrations.WebClient/MapExceptionCallWrapper.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Net;
+using System.Reflection;
+using System.Threading.Tasks;
+using Lykke.Blockchains.Integrations.WebClient.Exceptions;
+using Lykke.HttpClientGenerator.Infrastructure;
+using Refit;
+
+namespace Lykke.Blockchains.Integrations.WebClient
+{
+    internal class MapExceptionCallWrapper : ICallsWrapper
+    {
+        private readonly ConcurrentDictionary<MethodInfo, IExceptionMapper> _mappersCache;
+
+        public MapExceptionCallWrapper()
+        {
+            _mappersCache = new ConcurrentDictionary<MethodInfo, IExceptionMapper>();
+        }
+
+        public async Task<object> HandleMethodCall(MethodInfo targetMethod, object[] args, Func<Task<object>> innerHandler)
+        {
+            try
+            {
+                return await innerHandler();
+            }
+            catch (ApiException ex)
+            {
+                var mapper = _mappersCache.GetOrAdd(targetMethod, m =>
+                {
+                    var mapperAttribute = m.GetCustomAttribute<ExceptionMapperAttribute>();
+
+                    if (mapperAttribute == null)
+                    {
+                        return null;
+                    }
+
+                    return (IExceptionMapper) Activator.CreateInstance(mapperAttribute.MapperType);
+                });
+
+                mapper?.ThrowMappedExceptionOrPassThrough(ex);
+
+                switch (ex.StatusCode)
```

Comment:
What is the purpose of mapping exceptions generally one to one with StatusCode? We have "catch ... when" in C# 6+.

Reply:
We have additional mapper here. It will be specified by particular client to override default behaviour. Because of this we can't use "when".